File: src/main/java/algo/slidingWindow/ReplaceTheSubstringForBalancedString_1234.java
package algo.slidingWindow;

import java.util.HashMap;

public class ReplaceTheSubstringForBalancedString_1234 {
  public static int balancedString(String s) {
    // Find the shortest window so that when we remove the window from the string, the count of all
    // chars are less than total / 4
    var limit = s.length() / 4;
    var count = new HashMap<Character, Integer>();
    for (var i = 0; i < s.length(); i++) {
      count.put(s.charAt(i), count.getOrDefault(s.charAt(i), 0) + 1);
    }
    if (count.size() == 4 && count.values().stream().allMatch(v -> v >= limit)) {
      return 0;
    }

    int left = 0, right = 0;
    var minLen = Integer.MAX_VALUE;
    while (right < s.length()) {
      count.put(s.charAt(right), count.get(s.charAt(right)) - 1);
      if (count.values().stream().allMatch(v -> v <= limit)) {
        minLen = Math.min(minLen, right - left + 1);
        while (left < right) {
          minLen = Math.min(minLen, right - left + 1);
          count.put(s.charAt(left), count.get(s.charAt(left)) + 1);
          left++;
          if (count.values().stream().anyMatch(v -> v > limit)) {
            break;
          }
        }
      }
      right++;
    }

    return minLen;
  }

  public static void main(String[] args) {
    System.out.println(balancedString("QWER"));
    System.out.println(balancedString("QQWE"));
    System.out.println(balancedString("QQQW"));
    System.out.println(balancedString("WQWRQQQW"));
  }
}

File: src/main/kotlin/algo/slidingWindow/ReplaceTheSubstringForBalancedString_1234.kt
package algo.slidingWindow

class KtReplaceTheSubstringForBalancedString_1234 {
    // Find the shortest window so that when we remove the window from the string, the count of all
    // chars are less than total / 4
    fun balancedString(s: String): Int {
        val limit = s.length / 4
        val count: MutableMap<Char, Int> = mutableMapOf()
        s.forEach { ch ->
            count[ch] = count.computeIfAbsent(ch) {0} + 1
        }
        if (count.size == 4 && count.all { entry -> entry.value == limit }) {
            return 0
        }

        var left = 0; var right = 0
        var minLen = Int.MAX_VALUE
        while (right < s.length) {
            count[s[right]] = count[s[right]]!! - 1
            if (count.all { entry -> entry.value <= limit }) {
                minLen = minLen.coerceAtMost(right - left + 1)
                while (left < right) {
                    minLen = minLen.coerceAtMost(right - left + 1)
                    count[s[left]] = count[s[left]]!! + 1
                    left++
                    if (count.any() { entry -> entry.value > limit }) {
                        break
                    }
                }
            }
            right++
        }

        return minLen
    }
}

fun main() {
    val balance = KtReplaceTheSubstringForBalancedString_1234()
    println(balance.balancedString("QWER"))
    println(balance.balancedString("QQWE"))
    println(balance.balancedString("QQQW"))
}